test/end-to-end/utils.ts

```diff
@@ -15,16 +15,293 @@
+    .up(['core']);
+};
+
+const waitFor = <T>({
+  check,
+  validate,
+  timeout = DEFAULT_TIMEOUT,
+  interval = 1000,
+  timeoutMessage,
+  waitingMessage,
+}: {
+  check: () => Promise<T> | T;
+  validate: (result: T) => boolean;
+  timeout?: number;
+  interval?: number;
+  timeoutMessage: string;
+  waitingMessage?: string;
+}): Promise<T> => {
+  return new Promise((resolve, reject) => {
+    const startTime = Date.now();
+
+    const checkCondition = async () => {
+      try {
+        const result = await check();
+        if (validate(result)) {
+          resolve(result);
+          return;
+        }
+
+        if (waitingMessage !== undefined) {
+          console.log(waitingMessage);
+        }
+
+        if (Date.now() - startTime >= timeout) {
+          reject(new Error(timeoutMessage));
+          return;
+        }
+
+        setTimeout(checkCondition, interval);
+      } catch (error) {
+        reject(error);
+      }
+    };
+
+    checkCondition();
+  });
+};
+
+export const waitForBlocks = ({
+  coreDb,
+  stopHeight,
+  timeout,
+  checkInterval: interval,
+}: {
+  coreDb: Database;
+  stopHeight: number;
+  timeout?: number;
+  checkInterval?: number;
+}) => {
+  return waitFor({
+    check: () => getMaxHeight(coreDb)['MAX(height)'],
+    validate: (height) => height === stopHeight,
+    timeout,
+    interval,
+    timeoutMessage: `Timeout waiting for blocks to reach height ${stopHeight}`,
+    waitingMessage: `Waiting for blocks to import... Current height: ${getMaxHeight(coreDb)['MAX(height)']}, Target: ${stopHeight}`,
+  });
+};
+
+export const waitForLogMessage = ({
+  container,
+  expectedMessage,
+  timeout = DEFAULT_TIMEOUT,
+}: {
+  container: StartedGenericContainer;
+  expectedMessage: string;
+  timeout?: number;
+}) => {
+  return new Promise((resolve, reject) => {
+    const timeoutId = setTimeout(() => {
+      reject(new Error(`Timeout waiting for message: ${expectedMessage}`));
+    }, timeout);
+
+    container
+      .logs()
+      .then((logStream) => {
+        logStream.on('data', (data) => {
+          const log = data.toString('utf8');
+          if (log.includes(expectedMessage)) {
+            clearTimeout(timeoutId);
+            resolve(true);
+          }
+        });
+
+        logStream.on('error', (error) => {
+          clearTimeout(timeoutId);
+          reject(error);
+        });
+      })
+      .catch((error) => {
+        clearTimeout(timeoutId);
+        reject(error);
+      });
+  });
+};
+
+export const waitForBundleToBeIndexed = ({
+  id,
+  host = 'http://localhost:4000',
+  secret = 'secret',
+  timeout,
+  checkInterval: interval,
+}: {
+  id: string;
+  host?: string;
+  secret?: string;
+  timeout?: number;
+  checkInterval?: number;
+}) => {
+  return waitFor({
+    check: () => getBundleStatus({ id, host, secret }),
+    validate: (bundleStatus) =>
+      bundleStatus?.lastUnbundledAt !== null &&
+      bundleStatus?.dataItemCount !== null,
```

Comment:
I think the configuration option to reduce the interval and enable using it is on another branch right now, but we might want to wait for `lastFullyIndexedAt` here.

Reply:
lastFullyIndexedAt is just updated when update bundle timestamps run. Unless we change the update bundle timestamps to something like 1 seconds we will not have lastFullyIndexedAt populated. Maybe we should do that but I thought that lastUnbundledAt and dataItemCount would be enough.
What do you think about that?